admin-dev/themes/new-theme/js/components/form/form-object-mapper.js

```diff
@@ -0,0 +1,282 @@
+    // This event is emitted each time an object field is updated (from both input change and external event)
+    this.modelFieldUpdatedEventName = inputConfig.modelFieldUpdated || 'modelFieldUpdated';
+
+    this.modelMapping = {};
+    this.formMapping = {};
+    this.object = {};
+
+    this.initFormMapping();
+    this.watchUpdates();
+    this.updateFullObject();
+  }
+
+  /**
+   * Returns the object mapped to the form (current live state)
+   *
+   * @returns {*|{}}
+   */
+  getObject() {
+    return this.object;
+  }
+
+  /**
+   * Get a field from the object based on the model key
+   *
+   * @param modelKey (string)
+   *
+   * @returns {*|{}}
+   */
+  get(modelKey) {
+    const modelKeys = modelKey.split('.');
+
+    return $.serializeJSON.deepGet(this.object, modelKeys);
+  }
+
+  /**
+   * Returns the input associated to a model field (in case the field is mapped to
+   * several inputs the default one, first configured, is always used).
+   *
+   * @param modelKey {string}
+   * @returns {undefined|jQuery}
+   */
+  getInput(modelKey) {
+    if (!Object.prototype.hasOwnProperty.call(this.modelMapping, modelKey)) {
+      return undefined;
+    }
+
+    return $(`[name="${this.modelMapping[modelKey]}"]`);
+  }
+
+  /**
+   * Set a value to a field of the object based on the model key, the object itself is updated
+   * of course but the mapped inputs are also synced (all of them if multiple). Events are also
+   * triggered to indicate the object has been updated (the general and the individual field ones).
+   *
+   * @param modelKey {string}
+   * @param value {*|{}}
+   */
+  set(modelKey, value) {
+    if (!Object.prototype.hasOwnProperty.call(this.modelMapping, modelKey) || value === this.get(modelKey)) {
+      return;
+    }
+
+    // First update the inputs then the model, so that the event is sent at last
+    this.updateInputValue(modelKey, value);
+    this.updateObjectByKey(modelKey, value);
+    this.eventEmitter.emit(this.modelUpdatedEventName, this.object);
+  }
+
+  /**
+   * Watches if changes happens from the form or via an event.
+   */
+  watchUpdates() {
+    this.$form.on('keyup change dp.change', ':input', _.debounce(
+      (event) => this.inputUpdated(event),
+      200,
+      {maxWait: 1000},
+    ));
+    this.eventEmitter.on(this.updateModelEventName, () => this.updateFullObject());
+  }
+
+  /**
+   * Triggered when a form input has been changed.
+   *
+   * @param event {Object}
+   */
+  inputUpdated(event) {
+    const target = event.currentTarget;
+
+    if (!Object.prototype.hasOwnProperty.call(this.formMapping, target.name)) {
+      return;
+    }
+
+    const updatedValue = $(target).val();
+    const updatedModelKey = this.formMapping[target.name];
+
+    // Update the mapped input fields
+    this.updateInputValue(updatedModelKey, updatedValue);
+
+    // Then update model and emit event
+    this.updateObjectByKey(updatedModelKey, updatedValue);
```

Comment:
When I read this, I was like "Maybe you could have some situations where updateInputValue fail but updateObjectByKey would success or the invert", maybe it looks impossible but I just wanted to say it

Reply:
That's why I check if it matches the known mapping So if the mapping is correct both are supposed to work But I can add some `console.error` in each functions I guess (as we discussed) So that it can at least help during development time

Reply:
Maybe this is a good solution, if everyone agree on having console.error in our codebase (and it's removed on production, I think we should abuse on this)